Compiler primitives and standard library support for 63-bit integers

mplamann opened the pull request:
This patch adds an `int63` type to the OCaml compiler and standard library. Similar to `int32` and `int64`, `int63` represents a 63-bit integer that works on 32-bit and 64-bit systems. On 32-bit systems, an `int63` is represented as an `int64` with the bottom bit set to 1.  On 64-bit systems, an `int63` is represented as a standard OCaml `int`.

This allows the same code to use unboxed integers on 64-bit systems, while keeping the same bit width on 32-bit systems.

Tests are included in testsuite/tests/basic/test_int63.ml

m-plamann committed:
Rewrote stdlib to use int63 type rather than Int63.t

m-plamann committed:
Added toplevel printer for Int63

m-plamann committed:
Bootstrapped compiler with int63 primitives

m-plamann committed:
Added Int63 module to standard library

m-plamann committed:
Added Int63 tests

m-plamann committed:
Added int63 primtives to the compiler

m-plamann committed:
Update documentation for PR

m-plamann committed:
Added int63 type to predef.ml, updated bootstrap compiler to have int63 type

yallop commented:
The `caml_int63_ops` appear to be unused, since `int63` values are represented as `int64` or `int`.  If you do decide to use them, they need to be fixed: they currently use `Long_val` to access `int63` values, which won't work as expected on 32-bit architectures.

The implementations of `div` and `mod` are wrong.  What happens when the divisor is zero?  You need to do what the implementations for `int64` etc. do..

Why are `int63` values printed with a `t` suffix (`oprint.ml`) when that suffix is not otherwise supported (e.g. for literals)?

m-plamann committed:
Int63 bugfixes in response to PR comments

mplamann commented:
`caml_int63_ops` was dead code from a previous attempt at implementing int63, I've removed it.

`div` and `mod` should now raise an OCaml exception when dividing by zero, and I've added a test to test_int63.ml to cover this case.

`int63` values were printed with  a `t` suffix because I had started working on literal support for `int63`s. I wasn't able to implement `int63` literals that generated cross-platform bytecode, so I've left them out of this PR. Does it make sense to print `int63` values as `Int63.of_int64 1234L` at the toplevel?

rneswold commented:
The C programmer in me thinks this is a cool way to write portable code that is automatically optimized on 64-bit systems. The OCaml-student in me wonders why we're exposing an implementation detail to the programmer.

When someone writes a GPU backend for OCaml which has 128-bit integers, will this break? Or will we need to add `Int127` to the standard library?

mplamann commented:
I don't think this exposes any more implementation details than are already exposed to the programmer. If you as a programmer are in a situation where you'd want to use an `int32`, `int63`, or `int64`, you are likely already aware that the standard OCaml `int` is 31 or 63 bit integer, depending on your target architecture.

If this was compiled for a 128-bit architecture, I think an `int63` would be boxed just like on a 32-bit platform, and so this code would still work.

bobzhang commented:
what's the overflow semantics

chambart commented on middle_end/semantics_of_primitives.ml:
```diff
@@ -84,7 +84,24 @@ let for_primitive (prim : Lambda.primitive) =
+  | Psubint63
+  | Pmulint63
+  | Pdivint63
+  | Pmodint63
```
mod and div have effects (division by zero). You should take care if the pr https://github.com/ocaml/ocaml/pull/702 is merged before this one to apply the same changes here.

chambart commented:
Those won't benefit from constant propagation and others without writing quite a lot of specific code: you should instead convert the primitives to int primitives (on 64bits platforms) when going from lambda to clambda (asmcomp/closure.ml) or to flambda (middle_end/closure_conversion.ml)

mshinwell commented:
@diml Can you do the work of finishing this one off?

yallop commented on Changes:
```diff
@@ -14,6 +14,10 @@ Next version (tbd):
 
+### Standard Library
+- GPR#770: Add 63-bit integer support to the standard library
+  (Mitchell Plamann, Jeremie Diminio, Leo White)
```
"Diminio" ↝ "Dimino", "Jeremie" ↝ "Jérémie"

mshinwell commented:
We're planning to come back and finish this pull request later in 2017.

stedolan commented:
How does this interact with marshalling? If I marshal an int63 on a 64-bit system, is it going to come out as an int63 when unmarshalled on a 32-bit system? AFAIK, this issue is why 64-bit OCaml boxes `int32`, instead of representing it inside a plain `int`.

ghost commented:
Basically int63 values don't support marshaling between different architectures, I don't see how to workaround this

ghost commented:
We talked about this more at Jane Street. This patch ends up being quite large and invasive. Additionally, we still don't have a good answer regarding the marshaling of int63 values. So in the interest of lowering the review burden of compiler developers and not adding new types to the stdlib that don't play well with marshal, we decided to abandon this PR.

Instead, we wrote a much smaller and simpler patch that allows us to get the same benefit but with an `Int63` module written outside of the compiler. We are currently testing it inside Jane Street, and if all goes well we will submit our patch upstream.
